---
title: "Zebrafish"
output: html_document
date: "2025-06-11"
---

# load libraries and data
```{r}

library(Seurat)
library(tidyverse)

zcm_nahia <- readRDS('/Users/marwanbakr/docs/kimlab/scRNA-seq/zebrafish_heart/embryo/zebrafish_cm_20250401.rds')
zccs_nahia <- readRDS('/Users/marwanbakr/docs/kimlab/scRNA-seq/zebrafish_heart/embryo/zebrafish_ccs_20250401.rds')

zcm_li <- readRDS('/Users/marwanbakr/docs/kimlab/scRNA-seq/zebrafish_ncomms/zebrafish_cm_adult_li.rds')
zccs_li <- readRDS('/Users/marwanbakr/docs/kimlab/scRNA-seq/zebrafish_ncomms/zebrafish_ccs_adult_li.rds')

zfin_human <- read_tsv("https://zfin.org/downloads/human_orthos.txt", 
                       col_names = c("ZFIN_ID", "Zebrafish_Symbol", "Zebrafish_Desc", 
                                     "Human_Symbol", "Human_Desc", "OMIM", 
                                     "EntrezID", "Alt_Entrez", "AA", "ZFIN_Pub"),
                       show_col_types = FALSE)

zfin_mouse <- read_tsv("https://zfin.org/downloads/mouse_orthos.txt", 
                       col_names = c("ZFIN_ID", "Zebrafish_Symbol", "Zebrafish_Desc", 
                                     "Mouse_Symbol", "Mouse_Desc", "OMIM", 
                                     "EntrezID", "Alt_Entrez", "AA", "ZFIN_Pub"),
                       show_col_types = FALSE)

# create lookup tables
dr_to_hs <- setNames(zfin_human$Human_Symbol, zfin_human$Zebrafish_Symbol)
dr_to_mm <- setNames(zfin_mouse$Mouse_Symbol, zfin_mouse$Zebrafish_Symbol)

dr_to_hs["isl1"] <- "ISL1"
dr_to_mm["isl1"] <- "Isl1"

# mouse to human lookup
mm_to_hs <- setNames(ortholog_df$hs, ortholog_df$mm)

```

# integration
```{r}

zcm_li$batch <- zcm_li$group
zcm_nahia$batch <- zcm_nahia$sample

merged <- merge(subset(zcm_li, idents=c('Ventricle', 'Atrium')), zcm_nahia) # no 'uninjured' from zcm_li

merged[["percent.mt"]] <- PercentageFeatureSet(merged, pattern = "^mt-")
merged <- subset(merged, nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 20) # same as embryo dataset

merged <- NormalizeData(merged)
merged <- FindVariableFeatures(merged)
merged <- ScaleData(merged, features=rownames(merged))
merged <- RunPCA(merged)

# integrate via reciprocal PCA projection
merged <- IntegrateLayers(object = merged, method = RPCAIntegration, orig.reduction = "pca", new.reduction = "rpca")

merged[["RNA"]] <- JoinLayers(merged[["RNA"]])

merged <- FindNeighbors(merged, reduction = "rpca", dims = 1:20)
merged <- RunUMAP(merged, dims = 1:20, reduction = "rpca")

merged <- FindClusters(merged, resolution = 0.3)

merged <- subset(merged, idents=c('4', '6', '9', '8'), invert=T) # smooth muscle- (acta2, cnn1b, tagln), endothelial- (kdrl, ramp2), fibroblast- (thbs1a, dcn, pdgfrb, col1a1a), immune-like (cxcl8a, cd74a, ccr9a, il1b, c1qc, c1qb) respectively

merged <- NormalizeData(merged)
merged <- FindVariableFeatures(merged)
merged <- ScaleData(merged, features=rownames(merged))
merged <- RunPCA(merged)
merged <- FindNeighbors(merged, reduction = "rpca", dims = 1:20)
merged <- RunUMAP(merged, dims = 1:20, reduction = "rpca")
merged <- FindClusters(merged, resolution=0.15)

merged <- SetIdent(merged, WhichCells(merged, idents=c('0', '1', '4')), 'vCM') #myh7l
merged <- SetIdent(merged, WhichCells(merged, idents=c('2', '3')), 'aCM') #myh6

merged <-
  SetIdent(merged,
           WhichCells(
             merged,
             idents = 'vCM',
             expression = batch %in% c('Ventricle', 'Atrium')
           ),
           'vCM (6-12mo)')
merged <-
  SetIdent(merged,
           WhichCells(
             merged,
             idents = 'vCM',
             expression = batch %in% c('et31_48h', 'et31_72h', 'et33_48h', 'et33_72h')
           ),
           'vCM (48-72hpf)')
merged <-
  SetIdent(merged,
           WhichCells(
             merged,
             idents = 'aCM',
             expression = batch %in% c('Ventricle', 'Atrium')
           ),
           'aCM (6-12mo)')
merged <-
  SetIdent(merged,
           WhichCells(
             merged,
             idents = 'aCM',
             expression = batch %in% c('et31_48h', 'et31_72h', 'et33_48h', 'et33_72h')
           ),
           'aCM (48-72hpf)')
merged <-
  SetIdent(
    merged,
    WhichCells(
      merged,
      expression = batch %in% c('et31_48h', 'et31_72h', 'et33_48h', 'et33_72h') &
        sub_ccs == 'Atrioventricular CMs'
    ),
    'AVC (48-72hpf)'
  )

merged <-
  SetIdent(
    merged,
    WhichCells(
      merged,
      expression = batch %in% c('Ventricle', 'Atrium') &
        sub_ccs == 'SA node'
    ),
    'SA node (6-12mo)'
  )
merged <-
  SetIdent(
    merged,
    WhichCells(
      merged,
      expression = batch %in% c('et31_48h', 'et31_72h', 'et33_48h', 'et33_72h') &
        sub_ccs == 'SA node'
    ),
    'SA node (48-72hpf)'
  )

merged$sub_ccs2 <- Idents(merged)

zccs <- merge(subset(merged, idents=c("48-72hpf - SA node", "48-72hpf - AVC")), subset(merged, idents="6-12mo - SA node"))

zccs <- NormalizeData(zccs)
zccs <- FindVariableFeatures(zccs)
zccs <- ScaleData(zccs, features=rownames(zccs))
zccs <- RunPCA(zccs)

zccs <- IntegrateLayers(object = zccs, method = RPCAIntegration, orig.reduction = "pca", new.reduction = "rpca", k.weight=83)

zccs <- FindNeighbors(zccs, dims=1:20, reduction = 'rpca')
zccs <- RunUMAP(zccs, dims=1:20, reduction='rpca')

fig <- DimPlot(merged, pt.size=1.3) +
  scale_color_manual(
    values = c(
      "SA node (48-72hpf)" = "#700000",
      "SA node (6-12mo)" = "#C04A4A",
      "AVC (48-72hpf)" = "#e7ad0d",
      "aCM (48-72hpf)" = "#D5B09A",
      "aCM (6-12mo)" = "#ede0d4",
      "vCM (48-72hpf)" = "#8a817c",
      "vCM (6-12mo)" = "#bcb8b1"
    ),
    labels = c(
      "SA node (48-72hpf)",
      "SA node (6-12mo)",
      "AVC (48-72hpf)",
      "aCM (48-72hpf)",
      "aCM (6-12mo)",
      "vCM (48-72hpf)", 
      "vCM (6-12mo)"
    )
  ) +
  theme(
    legend.text = element_markdown(size = 22)
  ) +
  NoAxes()

ggsave('dr_integrated_umap.tiff', fig, dpi=600, height=7, width=12)

Idents(zccs) <- factor(Idents(zccs), levels=c('48-72hpf - SA node', '6-12mo - SA node', '48-72hpf - AVC'))

fig <- DimPlot(zccs, cols=c("#700000", "#C04A4A", "#e7ad0d"), pt.size=3) & NoAxes() & NoLegend()

ggsave('dr_integrated_umap_ccs_only.tiff', fig, dpi=600, height=7, width=7)

fig <-
  VlnPlot(
    merged,
    c(
      'shox2',
      'vsnl1a',
      'tbx2a',
      'tbx2b',
      'cx36.7',
      'myh7l',
      'myh6'
    ),
    idents = c('48-72hpf - SA node', '6-12mo - SA node', '48-72hpf - AVC', '48-72hpf - aCM', '6-12mo - aCM'),
    stack = T,
    flip = T,
    fill.by = 'ident'
    ) &
  stat_summary(
    fun = mean,
    geom = 'crossbar',
    width = 0.1,
    linewidth = 0.25,
    colour = "black"
  ) &
  theme(
    axis.title.x = element_blank(),
    axis.ticks.length.x = unit(0.1, "cm"),
    axis.ticks.length.y = unit(0.1, "cm"),
    strip.text = element_text(hjust = 0, face = "italic", size=16),
    axis.text.x = element_text(size=16)
  ) &
  NoLegend()

```

# separate objects
```{r}

Idents(merged) <- merged$batch

zcm_e <- subset(merged, idents=c('et31_48h', 'et31_72h', 'et33_48h', 'et33_72h'))
zcm_a <- subset(merged, idents=c('Atrium', 'Ventricle'))

zcm_e <- NormalizeData(zcm_e)
zcm_e <- FindVariableFeatures(zcm_e)
zcm_e <- ScaleData(zcm_e, features=rownames(zcm_e))
zcm_e <- RunPCA(zcm_e)
zcm_e <- FindNeighbors(zcm_e, dims=1:20)
zcm_e <- RunUMAP(zcm_e, dims=1:20)


zcm_a <- NormalizeData(zcm_a)
zcm_a <- FindVariableFeatures(zcm_a)
zcm_a <- ScaleData(zcm_a, features=rownames(zcm_a))
zcm_a <- RunPCA(zcm_a)
zcm_a <- FindNeighbors(zcm_a, dims=1:20)
zcm_a <- RunUMAP(zcm_a, dims=1:20)

```

# correlation between cell types (9-15pcw and 19pcw)
```{r}

# get pseudobulk matrices (scaled)
rownames(zcm_e) <- gsub("_", "-", rownames(zcm_e))
rownames(zcm_e) <- make.unique(rownames(zcm_e))

rownames(zcm_a) <- gsub("_", "-", rownames(zcm_a))
rownames(zcm_a) <- make.unique(rownames(zcm_a))

Idents(zcm_e) <- zcm_e$sub_ccs2
Idents(zcm_a) <- zcm_a$sub_ccs2

eavg <- AverageExpression(zcm_e, return.seurat=T)
eavg <- GetAssayData(eavg, assay = "RNA", slot = "scale.data")
aavg <- AverageExpression(zcm_a, return.seurat=T)
aavg <- GetAssayData(aavg, assay = "RNA", slot = "scale.data")

# subset to genes common to both datasets
common_genes <- intersect(rownames(eavg), rownames(aavg))
eavg <- eavg[common_genes, ]
aavg <- aavg[common_genes, ]

aavg <- aavg[, colnames(aavg) != "g6-12mo - aCM (bmp4+)"]

# compute correlation matrix
cor_mtx <- cor(eavg, aavg, method = "pearson")

library(pheatmap)

tiff(
  'PLACEHOLDER.tiff',
  width = 7,
  height = 6,
  units = "cm",
  res = 600
)

pheatmap(cor_mtx, cluster_rows = FALSE, cluster_cols = FALSE,
         color = colorRampPalette(c('#f5f9fe', '#7ab6d9', '#08326e'))(1000),
         border_color = NA, scale = "row", fontsize = 5)

dev.off()

```

```{r}

mark_san_dr_e <- FindMarkers(zcm_e, ident.1='48-72hpf - SA node', ident.2='48-72hpf - aCM', only.pos=T, min.pct=0.1)
mark_san_dr_e <- mark_san_dr_e %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1) # for FC > 1.5

mark_avn_dr_e <- FindMarkers(zcm_e, ident.1='48-72hpf - AVC', ident.2='48-72hpf - vCM', only.pos=T, min.pct=0.1)
mark_avn_dr_e <- mark_avn_dr_e %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1) # for FC > 1.5

dr_san_vs_avn_e <- FindMarkers(zcm_e, ident.1='48-72hpf - SA node', ident.2='48-72hpf - AVC', only.pos=T, min.pct=0.1)
dr_san_vs_avn_e <- dr_san_vs_avn_e %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)

dr_avn_vs_san_e <- FindMarkers(zcm_e, ident.1='48-72hpf - AVC', ident.2='48-72hpf - SA node', only.pos=T, min.pct=0.1)
dr_avn_vs_san_e <- dr_avn_vs_san_e %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)

mark_nodes_dr_e <- FindMarkers(zcm_e, ident.1 = c('48-72hpf - SA node', '48-72hpf - AVC'), ident.2 = c('48-72hpf - aCM', '48-72hpf - vCM'), only.pos = TRUE, min.pct = 0.1)
mark_nodes_dr_e <- mark_nodes_dr_e %>%
  filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)

dr_san_e <- intersect(rownames(mark_san_dr_e), rownames(dr_san_vs_avn_e))
dr_avn_e <- intersect(rownames(mark_avn_dr_e), rownames(dr_avn_vs_san_e))
dr_nodes_e <- rownames(mark_nodes_dr_e)

dr_nodes_e["gjd6"] <- "cx36.7" # for consistent naming

dr_san_to_hs_e <- dr_to_hs[dr_san_e]
dr_avn_to_hs_e <- dr_to_hs[dr_avn_e]
dr_nodes_to_hs_e <- dr_to_hs[dr_nodes_e]

mme_san_to_hs <- mm_to_hs[mme_san]
mme_avn_to_hs <- mm_to_hs[mme_avn]
mme_nodes_to_hs <- mm_to_hs[mme_nodes]

shared_san_hs_dr_e <- intersect(hse_san, dr_san_to_hs_e)
shared_avn_hs_dr_e <- intersect(hse_avn, dr_avn_to_hs_e)
shared_nodes_hs_dr_e <- intersect(hse_nodes, dr_nodes_to_hs_e)

shared_hs_san_mm_dr_e <- intersect(mme_san_to_hs, dr_san_to_hs_e)
shared_hs_avn_mm_dr_e <- intersect(mme_avn_to_hs, dr_avn_to_hs_e)
shared_hs_nodes_mm_dr_e <- intersect(mme_nodes_to_hs, dr_nodes_to_hs_e)

shared_san_mm_dr_e <- shared_hs_san_mm_dr_e[!is.na(shared_hs_san_mm_dr_e)]
shared_avn_mm_dr_e <- shared_hs_avn_mm_dr_e[!is.na(shared_hs_avn_mm_dr_e)]
shared_nodes_mm_dr_e <- shared_hs_nodes_mm_dr_e[!is.na(shared_hs_nodes_mm_dr_e)]

# Reverse map: human → all zebrafish genes (to handle one-to-many orthologs)
hs_to_dr_list <- split(names(dr_to_hs), dr_to_hs)

# Keep only valid human orthologs (non-NA)
valid_dr_san_hs <- dr_to_hs[dr_san_e]; valid_dr_san_hs <- valid_dr_san_hs[!is.na(valid_dr_san_hs)]
valid_dr_avn_hs <- dr_to_hs[dr_avn_e]; valid_dr_avn_hs <- valid_dr_avn_hs[!is.na(valid_dr_avn_hs)]
valid_dr_nodes_hs <- dr_to_hs[dr_nodes_e]; valid_dr_nodes_hs <- valid_dr_nodes_hs[!is.na(valid_dr_nodes_hs)]

# Zebrafish genes whose human orthologs are shared with human (hse_san)
shared_san_zf_dr_e_hs <- names(valid_dr_san_hs)[valid_dr_san_hs %in% hse_san]
shared_avn_zf_dr_e_hs <- names(valid_dr_avn_hs)[valid_dr_avn_hs %in% hse_avn]
shared_nodes_zf_dr_e_hs <- names(valid_dr_nodes_hs)[valid_dr_nodes_hs %in% hse_nodes]

# Zebrafish genes whose human orthologs are shared with mouse (mme_san_to_hs)
shared_san_zf_dr_e_mm <- names(valid_dr_san_hs)[valid_dr_san_hs %in% mme_san_to_hs]
shared_avn_zf_dr_e_mm <- names(valid_dr_avn_hs)[valid_dr_avn_hs %in% mme_avn_to_hs]
shared_nodes_zf_dr_e_mm <- names(valid_dr_nodes_hs)[valid_dr_nodes_hs %in% mme_nodes_to_hs]


```

# adult markers
```{r}

mark_san_dr_a <- FindMarkers(zcm_a, ident.1='6-12mo - SA node', ident.2='6-12mo - aCM', only.pos=T, min.pct=0.1)
mark_san_dr_a <- mark_san_dr_a %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1) # for FC > 1.5

dr_san_a <- rownames(mark_san_dr_a)

dr_san_to_hs_a <- dr_to_hs[dr_san_a]

mmp_san_to_hs <- mm_to_hs[mmp_san]

shared_san_hs_dr_a <- intersect(hsa_san, dr_san_to_hs_a)

shared_hs_san_mm_dr_a <- intersect(mmp_san_to_hs, dr_san_to_hs_a)

shared_san_mm_dr_a <- shared_hs_san_mm_dr_a[!is.na(shared_hs_san_mm_dr_a)]

# Reverse map: human → all zebrafish genes (to handle one-to-many orthologs)
hs_to_dr_list <- split(names(dr_to_hs), dr_to_hs)

# Keep only valid human orthologs (non-NA)
valid_dr_san_hs <- dr_to_hs[dr_san_a]; valid_dr_san_hs <- valid_dr_san_hs[!is.na(valid_dr_san_hs)]

shared_san_zf_dr_a_hs <- names(valid_dr_san_hs)[valid_dr_san_hs %in% hsa_san]

shared_san_zf_dr_a_mm <- names(valid_dr_san_hs)[valid_dr_san_hs %in% mmp_san_to_hs]


```

# shared between zebrafish stages
```{r}

shared_san_dr <- intersect(dr_san_e, dr_san_a)
shared_san_dr_only <- setdiff(shared_san_dr, c(shared_san_zf_dr_e_mm, shared_san_zf_dr_e_hs, shared_san_zf_dr_a_mm, shared_san_zf_dr_a_hs))

```

# zebrafish embryo specific
```{r}

library(scCustomize)

# SAN

# get % expression of embryo SAN markers in all other datasets
percent_zf_a_san <- Percent_Expressing(zcm_a, dr_san_e, group_by = "sub_ccs2")
percent_mm_e_san <- Percent_Expressing(gccs, ortholog_df_zf$mm[ortholog_df_zf$dr %in% dr_san_e], group_by = "sub_ccs")
percent_hs_e_san <- Percent_Expressing(fccs, ortholog_df_zf$hs[ortholog_df_zf$dr %in% dr_san_e], group_by = "sub_ccs")

# get low expression genes in each species
hs_low_expr <- rownames(percent_hs_e_san[percent_hs_e_san$SAN.tail < 5 & percent_hs_e_san$SAN.head < 5, ])
mm_low_expr <- rownames(percent_mm_e_san[percent_mm_e_san$SAN.head < 5 & percent_mm_e_san$SAN.tail < 5, ])
zf_a_low_expr <- rownames(percent_zf_a_san[percent_zf_a_san$X6.12mo...SA.node < 5, ])

# map human/mouse low expr to zebrafish genes
hs_low_expr_dr <- ortholog_df_zf$dr[ortholog_df_zf$hs %in% hs_low_expr]
mm_low_expr_dr <- ortholog_df_zf$dr[ortholog_df_zf$mm %in% mm_low_expr]

# filter to SAN list
hs_low_expr_dr <- intersect(hs_low_expr_dr, dr_san_e)
mm_low_expr_dr <- intersect(mm_low_expr_dr, dr_san_e)
zf_a_low_expr <- intersect(zf_a_low_expr, dr_san_e)

# final zebrafish embryo-specific SAN list
zf_san_specific_e <- intersect(dr_san_e, intersect(hs_low_expr_dr, intersect(mm_low_expr_dr, zf_a_low_expr)))

# AVN

percent_mm_e_avn <- Percent_Expressing(gccs, ortholog_df_zf$mm[ortholog_df_zf$dr %in% dr_avn_e], group_by = "sub_ccs")
percent_hs_e_avn <- Percent_Expressing(fccs, ortholog_df_zf$hs[ortholog_df_zf$dr %in% dr_avn_e], group_by = "sub_ccs")

hs_low_expr_avn <- rownames(percent_hs_e_avn[percent_hs_e_avn$Compact.AVN < 5, ])
mm_low_expr_avn <- rownames(percent_mm_e_avn[percent_mm_e_avn$Compact.AVN < 5, ])

hs_low_expr_dr_avn <- intersect(ortholog_df_zf$dr[ortholog_df_zf$hs %in% hs_low_expr_avn], dr_avn_e)
mm_low_expr_dr_avn <- intersect(ortholog_df_zf$dr[ortholog_df_zf$mm %in% mm_low_expr_avn], dr_avn_e)


zf_avn_specific_e <- intersect(dr_avn_e, intersect(hs_low_expr_dr_avn, mm_low_expr_dr_avn))


# nodes (SAN + AVC)

percent_mm_e_nodes <- Percent_Expressing(gccs, ortholog_df_zf$mm[ortholog_df_zf$dr %in% dr_nodes_e], group_by = "sub_ccs")
percent_hs_e_nodes <- Percent_Expressing(fccs, ortholog_df_zf$hs[ortholog_df_zf$dr %in% dr_nodes_e], group_by = "sub_ccs")

hs_low_expr_nodes <- rownames(percent_hs_e_nodes[
  percent_hs_e_nodes$SAN.tail < 5 &
  percent_hs_e_nodes$SAN.head < 5 &
  percent_hs_e_nodes$Compact.AVN < 5, ])

mm_low_expr_nodes <- rownames(percent_mm_e_nodes[
  percent_mm_e_nodes$SAN.head < 5 &
  percent_mm_e_nodes$SAN.tail < 5 &
  percent_mm_e_nodes$Compact.AVN < 5, ])

hs_low_expr_dr_nodes <- intersect(ortholog_df_zf$dr[ortholog_df_zf$hs %in% hs_low_expr_nodes], dr_nodes_e)
mm_low_expr_dr_nodes <- intersect(ortholog_df_zf$dr[ortholog_df_zf$mm %in% mm_low_expr_nodes], dr_nodes_e)


zf_nodes_specific_e <- intersect(dr_nodes_e, intersect(hs_low_expr_dr_nodes, mm_low_expr_dr_nodes))

```

# zebrafish adult specific
```{r}

library(scCustomize)

# SAN 

# get % expression of embryo SAN markers in all other datasets
percent_zf_e_san <- Percent_Expressing(zcm_e, dr_san_a, group_by = "sub_ccs2")
percent_mm_a_san <- Percent_Expressing(occs, ortholog_df_zf$mm[ortholog_df_zf$dr %in% dr_san_a], group_by = "sub_ccs")
percent_hs_a_san <- Percent_Expressing(kccs, ortholog_df_zf$hs[ortholog_df_zf$dr %in% dr_san_a], group_by = "sub_ccs")

# get low expression genes in each species
hs_low_expr_san <- rownames(percent_hs_a_san[percent_hs_a_san$SA.node < 5, ])
mm_low_expr_san <- rownames(percent_mm_a_san[percent_mm_a_san$SA.node < 5, ])
zf_e_low_expr_san <- rownames(percent_zf_e_san[percent_zf_e_san$X48.72hpf...SA.node < 5, ])

# map human/mouse low expr to zebrafish genes
hs_low_expr_dr_san <- ortholog_df_zf$dr[ortholog_df_zf$hs %in% hs_low_expr_san]
mm_low_expr_dr_san <- ortholog_df_zf$dr[ortholog_df_zf$mm %in% mm_low_expr_san]

# filter to SAN list
hs_low_expr_dr_san <- intersect(hs_low_expr_dr_san, dr_san_a)
mm_low_expr_dr_san <- intersect(mm_low_expr_dr_san, dr_san_a)
zf_e_low_expr_san <- intersect(zf_e_low_expr_san, dr_san_a)

# final zebrafish embryo-specific SAN list
zf_san_specific_a <- intersect(dr_san_a, intersect(hs_low_expr_dr_san, intersect(mm_low_expr_dr_san, zf_e_low_expr_san)))


```

# zebrafish and medaka trabeculae vs mouse/human PF
```{r}

# from Carey et al., Biol. Open

mc <- readRDS('/Users/marwanbakr/docs/kimlab/scRNA-seq/zebrafish_medaka_heart/myocardium.rds')


# zebrafish: uninjured, trabecular vs cortical
zf_uninjured <- subset(mc, subset = group == "zebrafish_uninjured")
Idents(zf_uninjured) <- zf_uninjured$cell_class
zf_trbc <- FindMarkers(zf_uninjured, ident.1 = "Trabecular", ident.2 = "Cortical", only.pos=T, min.pct=0.1, assay = "RNA")
zf_trbc <-  zf_trbc %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)
zf_trbc <- rownames(zf_trbc)

zf_cort <- FindMarkers(zf_uninjured, ident.1 = "Cortical", ident.2 = "Trabecular", only.pos=T, min.pct=0.1, assay="RNA")
zf_cort <-  zf_cort %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)
zf_cort <- rownames(zf_cort)


# medaka: uninjured, trabecular vs cortical
mdk_uninjured <- subset(mc, subset = group == "medaka_uninjured")
Idents(mdk_uninjured) <- "cell_class"
mdk_trbc <- FindMarkers(mdk_uninjured, ident.1 = "Trabecular", ident.2 = "Cortical", only.pos=T, min.pct=0.1, assay="RNA")
mdk_trbc <-  mdk_trbc %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)
mdk_trbc <- rownames(mdk_trbc)

mdk_cort <- FindMarkers(mdk_uninjured, ident.1 = "Cortical", ident.2 = "Trabecular", only.pos=T, min.pct=0.1, assay="RNA")
mdk_cort <-  mdk_cort %>% filter(p_val_adj <= 0.05 & avg_log2FC >= log2(1.5) & pct.1 >= 0.1)
mdk_cort <- rownames(mdk_cort)

dr_to_hs <- c(dr_to_hs, scn5laa = "SCN5A", scn5lab = "SCN5A") # not in list, added manually

zf_trbc_to_hs <- dr_to_hs[zf_trbc]
zf_trbc_to_hs <- zf_trbc_to_hs[!is.na(zf_trbc_to_hs)]
mdk_trbc_to_hs <- dr_to_hs[mdk_trbc]
mdk_trbc_to_hs <- mdk_trbc_to_hs[!is.na(mdk_trbc_to_hs)]

library(ComplexUpset)
library(tidyverse)
library(scales)

zf_gene_lists <- list(
  zf_trbc = zf_trbc_to_hs,        # zebrafish trabeculae --> human orthologs
  hse_pf  = hse_pf,               # human fetal PF
  hsa_pf  = hsa_pf,               # human adult PF
  mme_pf  = mm_to_hs_pf_e$hs,     # mouse embryonic PF --> human orthologs
  mmp_pf  = mm_to_hs_pf_a$hs      # mouse postnatal  PF --> human orthologs
)

all_genes_zf <- unique(unlist(zf_gene_lists))
gene_df_zf   <- as.data.frame(sapply(zf_gene_lists, `%in%`, x = all_genes_zf))
gene_df_zf$gene <- all_genes_zf

# keep only genes present in zebrafish trabeculae
zf_df_filtered <- gene_df_zf %>% filter(zf_trbc)

fig <- upset(
  zf_df_filtered,
  intersect = rev(names(zf_gene_lists)),
  sort_sets = F,
  min_degree = 2,
  base_annotations = list(
    'Intersection size' =
      intersection_size(counts = FALSE) +
      scale_y_continuous(breaks = breaks_width(5)) +
      theme(panel.grid = element_blank())
  ),
  set_sizes = F,
  matrix = intersection_matrix(geom  = geom_point(size = 4)),
  height_ratio=2,
  stripes='white',
  sort_intersections_by = 'degree'
) +
  theme(axis.title.x = element_blank(), panel.grid = element_blank())

ggsave('zebrafish_trbc_mammal_vcs_upset.svg', fig, dpi=600, width=7, height=5)


# medaka gene lists

mdk_gene_lists <- list(
  mdk_trbc = mdk_trbc_to_hs,
  hse_pf  = hse_pf,
  hsa_pf  = hsa_pf,
  mme_pf  = mm_to_hs_pf_e$hs,
  mmp_pf  = mm_to_hs_pf_a$hs
)

all_genes_mdk <- unique(unlist(mdk_gene_lists))
gene_df_mdk   <- as.data.frame(sapply(mdk_gene_lists, `%in%`, x = all_genes_mdk))
gene_df_mdk$gene <- all_genes_mdk

# keep only genes present in medaka trabeculae
mdk_df_filtered <- gene_df_mdk %>% filter(mdk_trbc)

fig <- upset(
  mdk_df_filtered,
  intersect = rev(names(mdk_gene_lists)),
  sort_sets = F,
  width_ratio = 0.25,
  set_sizes = FALSE,
  min_degree= 2,
  base_annotations = list(
    'Intersection size' =
      intersection_size(counts = FALSE) +
      scale_y_continuous(breaks = breaks_width(5)) +
      theme(panel.grid = element_blank())
  ),
  matrix = intersection_matrix(geom  = geom_point(size = 4)),
  height_ratio=2,
  stripes='white'
) +
  theme(axis.title.x = element_blank(), panel.grid = element_blank())

ggsave('medaka_trbc_mammal_vcs_upset.svg', fig, dpi=600, width=7, height=5)

# generate lists of genes for each intersection combination (for reference)

library(dplyr)
library(tidyr)
library(stringr)

sets_mat <- zf_df_filtered %>%
  select(zf_trbc, hse_pf, hsa_pf, mme_pf, mmp_pf, gene)   # order matches plot

# derive a “combination label” for every gene row
combos_tbl <- sets_mat %>%
  rowwise() %>%
  mutate(combo = paste(names(.)[c_across(zf_trbc:mmp_pf)], collapse = " & ")) %>%
  ungroup()

# keep only combos that involve 2 or more sets  (same criterion as the upset plot)
combos_tbl <- combos_tbl %>%
  filter(str_count(combo, "&") + 1 >= 2)

# build human --> (all zf paralogs) mapping, filtering to only those in zf_trbc
hs_to_zf_multi <- split(names(zf_trbc_to_hs), zf_trbc_to_hs)

# filter paralogs to just those present in zf_trbc
hs_to_zf_filtered <- lapply(hs_to_zf_multi, function(paralogs) {
  intersect(paralogs, zf_trbc)
})

zf_order <- factor(zf_trbc, levels = zf_trbc)

genes_per_combo_zf <- combos_tbl %>%
  rowwise() %>%
  mutate(
    zf_paralogs = list(hs_to_zf_filtered[[gene]] %||% character(0))
  ) %>%
  ungroup() %>%
  group_by(combo) %>%
  summarise(
    n_genes = n(),
    genes = zf_paralogs %>%
      flatten_chr() %>%
      unique() %>%
      { .[order(match(., zf_order))] } %>%
      paste(collapse = ", ")
  ) %>%
  arrange(desc(n_genes))

genes_per_combo_zf <- genes_per_combo_zf %>%
  filter(combo != "zf_trbc & gene")

# same as above, but for medaka

sets_mat <- mdk_df_filtered %>%
  select(mdk_trbc, hse_pf, hsa_pf, mme_pf, mmp_pf, gene) # order matches plot

combos_tbl <- sets_mat %>%
  rowwise() %>%
  mutate(combo = paste(names(.)[c_across(mdk_trbc:mmp_pf)], collapse = " & ")) %>%
  ungroup()

combos_tbl <- combos_tbl %>%
  filter(str_count(combo, "&") + 1 >= 2)

hs_to_mdk_multi <- split(names(mdk_trbc_to_hs), mdk_trbc_to_hs)

hs_to_mdk_filtered <- lapply(hs_to_mdk_multi, function(paralogs) {
  intersect(paralogs, mdk_trbc)
})

mdk_order <- factor(mdk_trbc, levels = mdk_trbc)

genes_per_combo_mdk <- combos_tbl %>%
  rowwise() %>%
  mutate(
    mdk_paralogs = list(hs_to_mdk_filtered[[gene]] %||% character(0))
  ) %>%
  ungroup() %>%
  group_by(combo) %>%
  summarise(
    n_genes = n(),
    genes = mdk_paralogs %>%
      flatten_chr() %>%
      unique() %>%
      { .[order(match(., mdk_order))] } %>%
      paste(collapse = ", ")
  ) %>%
  arrange(desc(n_genes))

genes_per_combo_mdk <- genes_per_combo_mdk %>%
  filter(combo != "mdk_trbc & gene")


```

# umap with zebrafish and medaka trabeculae
```{r}

zfmdk <- merge(zf_uninjured, mdk_uninjured)

DefaultAssay(zfmdk) <- "RNA"
zfmdk <- NormalizeData(zfmdk)
zfmdk <- FindVariableFeatures(zfmdk)

anchors <- FindIntegrationAnchors(
  object.list = list(zf_uninjured, mdk_uninjured),
  normalization.method = "LogNormalize",
  reduction            = "rpca"
)
zfmdk <- IntegrateData(
  anchorset             = anchors,
  normalization.method  = "LogNormalize"
)

DefaultAssay(zfmdk) <- "integrated"

zfmdk <- ScaleData(zfmdk, features=rownames(zfmdk))
zfmdk <- RunPCA(zfmdk)
zfmdk <- FindNeighbors(zfmdk, dims=1:20)
zfmdk <- RunUMAP(zfmdk, dims=1:20)

zfmdk <- SetIdent(zfmdk, WhichCells(zfmdk, expression = species == 'medaka' & cell_class == 'Cortical'), 'Cortical (medaka)')
zfmdk <- SetIdent(zfmdk, WhichCells(zfmdk, expression = species == 'zebrafish' & cell_class == 'Cortical'), 'Cortical (zebrafish)')
zfmdk <- SetIdent(zfmdk, WhichCells(zfmdk, expression = species == 'medaka' & cell_class == 'Trabecular'), 'Trabeculae (medaka)')
zfmdk <- SetIdent(zfmdk, WhichCells(zfmdk, expression = species == 'zebrafish' & cell_class == 'Trabecular'), 'Trabeculae (zebrafish)')

fig <-DimPlot(zfmdk, cols = c('#1b4965', '#62b6cb', '#857D8C', '#CCC9CF'), pt.size=1.3) & NoAxes() & theme(legend.text = element_text(size = 22))

ggsave('/Users/marwanbakr/docs/kimlab/manuscripts/2024 CCS comparative/figures/fig7/zf_mdk_trbc_cort_umap.tiff', fig, dpi=600, height=7.5, width=12)

```