---
title: "Bulk_RNAseq_Comparison"
output: html_document
date: "2025-06-11"
---

# load libraries and data
```{r}

library(tidyverse)
library(Seurat)
library(VennDiagram)
library(scales)
library(readxl)
library(grid)

# from GSE65658 - Vedantham et al., Circ. Res. 2015
bulk_san_p4 <- read_excel('305913r1_online_table_ii.xlsx', sheet = "P4") 
bulk_san_p14 <- read_excel('305913r1_online_table_ii.xlsx', sheet = "P14")

san_enriched_p4 <- bulk_san_p4 %>%
  filter(`Log(2) Fold-Change RA vs SAN` < -log2(1.5), FDR < 0.05) %>%
  pull(Gene) %>%
  na.omit()

san_enriched_p14 <- bulk_san_p14 %>%
  filter(`Log(2) Fold-Change RA vs SAN` < -log2(1.5), FDR < 0.05) %>%
  pull(Gene) %>%
  na.omit()

# from GSE291597 - Hillestad et al., Jacc Clinical Electrophysiology 2025
bulk_pf <- read_excel('/Users/marwanbakr/docs/kimlab/manuscripts/2024 CCS comparative/figures/supp_figs/supp_fig_bulkrna/GSE291597_DEG_PF_vs_EC_VM_DESeq2.xlsx')

pf_enriched_bulk <- bulk_pf %>%
  filter(
    log2FC_PF.vs.VM > log2(1.5),
    as.numeric(padjPF.vs.VM) < 0.05
  ) %>%
  pull(GeneName) %>%
  na.omit

```

# venn diagrams for comparison
```{r}

v <- venn.diagram(
  x = list(
    "Bulk P4" = san_enriched_p4,
    "Bulk P14" = san_enriched_p14,
    "Single-cell P4" = mmp_san
  ),
  filename = NULL,
  height = 3000,
  width = 3000,
  resolution = 800,
  lwd = 1.5,
  col = c("#26547c", "#ef476f", "#ffd166"),
  fill = c(alpha("#26547c", 0.3), alpha("#ef476f", 0.3), alpha("#ffd166", 0.3)),
  cex = 1,
  fontfamily = "sans",
  cat.cex = 0,
  margin = 0.05
)

svg("sc_bulk_P4_P14_SAN_venn.svg", width = 3.75, height = 3.75, family = "sans")
grid.draw(v)
dev.off()

v <- venn.diagram(
  x = list(
    "Bulk PF" = pf_enriched_bulk,
    "Single-cell PF" = hsa_pf
  ),
  filename = NULL,
  height = 3000,
  width = 3000,
  resolution = 800,
  lwd = 1.5,
  col = c("#26547c", "#ef476f"),
  fill = c(alpha("#26547c", 0.3), alpha("#ef476f", 0.3)),
  cex = 1,
  fontfamily = "sans",
  cat.cex = 0,
  margin = 0.05
)

svg("sc_bulk_PF_venn.svg", width = 3.75, height = 3.75, family = "sans")
grid.draw(v)
dev.off()

```

# generate UpSet plots showing intersections of bulk RNA-seq with other species/stages sc/snRNA-seq

# SA node
```{r}

library(ComplexUpset)

# combine P4 and P14 gene lists
bulk_union <- union(
  ortholog_df[ortholog_df$mm %in% san_enriched_p4, ]$hs,
  ortholog_df[ortholog_df$mm %in% san_enriched_p14, ]$hs
)
bulk_filtered <- setdiff(bulk_union, mm_to_hs_san_a$hs)

san_gene_lists <- list(
  bulk_san = bulk_filtered,
  mme_san  = mm_to_hs_san_e$hs,
  hse_san  = hse_san,
  hsa_san  = hsa_san
)

# build binary matrix of gene presence
all_san_genes <- unique(unlist(san_gene_lists))
gene_df_san <- as.data.frame(sapply(san_gene_lists, `%in%`, x = all_san_genes))
gene_df_san$gene <- all_san_genes

san_df_filtered <- gene_df_san %>%
  filter(
    gene %in% bulk_filtered,        # filter genes present in the bulk set
    rowSums(select(., -gene)) >= 2  # present in 2 or more sets
  )

fig <- upset(
  san_df_filtered,
  intersect = rev(names(san_gene_lists)),
  sort_sets = FALSE,
  min_degree = 2,
  base_annotations = list(
    'Intersection size' =
      intersection_size(counts = FALSE) +
      scale_y_continuous(breaks = breaks_width(5)) +
      theme(panel.grid = element_blank())
  ),
  set_sizes = FALSE,
  matrix = intersection_matrix(geom = geom_point(size = 4)),
  height_ratio = 2,
  stripes = 'white',
  sort_intersections_by = 'degree'
) +
  theme(axis.title.x = element_blank(), panel.grid = element_blank())

ggsave('san_upset.svg', fig, dpi = 600, width = 7, height = 5)

long_df <- san_df_filtered %>%
  pivot_longer(
    cols = -gene,
    names_to = "set",
    values_to = "present"
  ) %>%
  filter(present)  # keep only present entries

collapsed <- long_df %>%
  group_by(gene) %>%
  summarize(intersection = paste(sort(set), collapse = "&"), .groups = "drop")

intersections_df <- collapsed %>%
  group_by(intersection) %>%
  summarise(genes = list(gene), count = n(), .groups = "drop")



```

# Purkinje fibers
```{r}

bulk_filtered_pf <- setdiff(pf_enriched_bulk, hsa_pf)  # exclude adult human PF from snRNA-seq

pf_gene_lists <- list(
  bulk_pf = bulk_filtered_pf,
  hse_pf  = hse_pf,
  mme_pf  = mm_to_hs_pf_e$hs,
  mmp_pf  = mm_to_hs_pf_a$hs
)

all_pf_genes <- unique(unlist(pf_gene_lists))
gene_df_pf <- as.data.frame(sapply(pf_gene_lists, `%in%`, x = all_pf_genes))
gene_df_pf$gene <- all_pf_genes

pf_df_filtered <- gene_df_pf %>%
  filter(
    gene %in% bulk_filtered_pf,
    rowSums(select(., -gene)) >= 2
  )

fig <- upset(
  pf_df_filtered,
  intersect = rev(names(pf_gene_lists)),
  sort_sets = FALSE,
  min_degree = 2,
  base_annotations = list(
    'Intersection size' =
      intersection_size(counts = FALSE) +
      scale_y_continuous(breaks = breaks_width(5)) +
      theme(panel.grid = element_blank())
  ),
  set_sizes = FALSE,
  matrix = intersection_matrix(geom = geom_point(size = 4)),
  height_ratio = 2,
  stripes = 'white',
  sort_intersections_by = 'degree'
) +
  theme(axis.title.x = element_blank(), panel.grid = element_blank())

ggsave('pf_upset.svg', fig, dpi = 600, width = 7, height = 5)

long_df_pf <- pf_df_filtered %>%
  pivot_longer(-gene, names_to = "set", values_to = "present") %>%
  filter(present)

collapsed_pf <- long_df_pf %>%
  group_by(gene) %>%
  summarize(intersection = paste(sort(set), collapse = "&"), .groups = "drop")

intersections_df_pf <- collapsed_pf %>%
  group_by(intersection) %>%
  summarise(genes = list(gene), count = n(), .groups = "drop")

```